### ISSUE TITLE
Support Swoole

### ISSUE THREAD

**Issue by tamer-hassan** [octobercms/october#3746]
Already got OctoberCMS working with Swoole after some work with https://github.com/scil/LaravelFly  running in "Map" mode.

but have only two main issues with current OC code:

- the stage at which Execution Context (back-end/front-end) setting happens is very early
See: https://github.com/octobercms/library/blob/master/src/Foundation/Bootstrap/RegisterOctober.php#L40-L50
It is not until System\ServiceProvider is registered that ExecutionContext is actually checked
See:
https://github.com/octobercms/october/blob/master/modules/system/ServiceProvider.php#L73
So why not make that check at that later stage?
This would allow us to run the bootstrapper's code only once, and only register the System ServiceProvider on every new request.

- Global vars usage is troublesome. Had a problem where all AJAX widgets were not working and traced it down to the use of global $_POST in `post()` helper function in:
https://github.com/octobercms/library/blob/master/src/Support/helpers.php#L42-L51
it was always coming back empty. however, `request()->post()` works for both php-fpm and Swoole.
also LaravelS mentions cannot use global vars like $_POST/$_GET etc: https://github.com/hhxsv5/laravel-s#important-notices

**Comment by tamer-hassan** [comment]
https://github.com/tamer-hassan/october-fly

**Comment by LukeTowers** [comment]
@tamer-hassan I don't really have any idea what you're asking for nor do I have the time to research a bunch of other projects to figure it out either. Try simplifying what you're saying so that I can easily grasp what it is you're trying to do / want me to do about your question.

**Comment by LukeTowers** [comment]
@tamer-hassan how about you submit a PR with the required changes to support Swoole to the https://gitub.com/octobercms/library repository and we can carry on the discussion there?

**Comment by tamer-hassan** [comment]
@LukeTowers PR's submitted, with changes as minimalist as possible.

1. Moving only the execution-context setting of back-end/front-end to System\ServiceProvider allows Swoole-based projects to work properly when only System\ServiceProvider is configured as a 'providers_in_request' (as in [OctoberFly example config](https://github.com/tamer-hassan/october-fly/blob/master/config/laravelfly-app-config.example.php) ), and everything else booted on worker (kept in memory). I have also tested the changes on php-fpm and it works as expected.

2. using request() helper instead of Globals $_GET / $_POST fixes issues with Swoole like inability to login, and also lets all AJAX based widgets work. Also tested on php-fpm an it all works as expected.

Also I've simplified OctoberFly itself a bit in 0.0.2, so that no editing of October Http Kernel is needed, while still allowing same code to run with both php-fpm & Swoole.

The performance gains are definitely good so far.

**Comment by LukeTowers** [comment]
@tamer-hassan I'm thinking that the accompanying plugin should probably live under the octoberrain vendor since we're supporting swoole within the core. Do you have any thoughts on that?

**Comment by tamer-hassan** [comment]
@LukeTowers october-fly requires `scil/laravel-fly`, which in turn requires `ext-swoole`.
https://github.com/scil/LaravelFly/blob/master/composer.json#L20
Swoole extension is only available for installation on Linux (via PECL) and mac OS X (via brew).

**Comment by LukeTowers** [comment]
Right, but I'm suggesting that October-Fly lives under https://github.com/octoberrain/swoole-plugin. That way support for Swoole within October is more of a first class citizen with a more centralized place for maintaining support.

**Comment by tamer-hassan** [comment]
Sounds good. Let's do that!

**Comment by w20k** [comment]
Better to ask @tamer-hassan ;)

**Comment by LukeTowers** [comment]
@Johnzero I'm not opposed to working on it but I don't have any use for it right now nor any funding to work on it.

**Comment by domstarkey** [comment]
@LukeTowers - I've been trying to contact you about exactly that :) 
I've sent you an email, just fyi

**Comment by tamer-hassan** [comment]
@LukeTowers I would like to continue maintaining october-fly, and there's obviously funding now available (per domstarkey's comment above). What's keeping us from moving forward?

**Comment by LukeTowers** [comment]
@tamer-hassan trying to coordinate a meeting time with @domstarkey, my life's pretty hectic right now.

**Comment by daftspunk** [comment]
> the stage at which Execution Context (back-end/front-end) setting happens is very early

Execution context now occurs from a provider and lives inside the container, see 0ae48e9f968bac97614838c14f16be6c619a6ae0

**Comment by daftspunk** [comment]
> Global vars usage is troublesome. Had a problem where all AJAX widgets were not working and traced it down to the use of global $_POST in post() helper function in:

Request values (GET/POST) are now pulled from the app container, see
https://github.com/octobercms/library/commit/fd839047c2d7757f225a8ed60673ca8a20fe7df4

The `header_remove()` call has been removed entirely because it did nothing:
https://github.com/octobercms/october/commit/c5bc804d7398522783fbd76b2764f9541747867d

There is no more premature termination of the application thread, no more `exit()` or `die()` used:
https://github.com/octobercms/october/commit/5190c8177bbb6531836ad8fa8458111361a35608

That should just about cover it. If there are any other issues with Swoole compatibility we should create a new issue for it and reference this one.

Thanks!
